sphinx --with-re2 fails to build due to no matching template for std::call_once

- [x] Ran `brew update` and retried your prior step?
- [x] Ran `brew doctor`, fixed as many issues as possible and retried your prior step?

----

[gist](https://gist.github.com/aw1621107/1561bc01b2cf4d0f45c13b7a811e86fc)

Looks like re2 is looking for the variadic overload of `std::call_once`:

```
template<class _Callable, class... _Args>
friend
void call_once(once_flag&, _Callable&&, _Args&&...);
```

But for some reason it isn't finding it:

```
In file included from sphinx.cpp:72:
/usr/local/include/re2/re2.h:908:5: error: no matching function for call to 'call_once'
    std::call_once(once_, &LazyRE2::Init, this);
    ^~~~~~~~~~~~~~
clang++ -DHAVE_CONFIG_H -I. -I../config    -DSYSCONFDIR="\"/usr/local/Cellar/sphinx/2.2.11/etc\"" -DDATADIR="\"/usr/local/var/data\"" -I/usr/local/include   -I../libstemmer_c/include     -fno-rtti -Wall -g -D_FILE_OFFSET_BITS=64 -O3 -DNDEBUG -c -o sphinxfilter.o sphinxfilter.cpp
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/mutex:557:1: note: candidate function template not viable: requires 2 arguments, but 3 were provided
call_once(once_flag& __flag, _Callable& __func)
^
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin/../include/c++/v1/mutex:569:1: note: candidate function template not viable: requires 2 arguments, but 3 were provided
call_once(once_flag& __flag, const _Callable& __func)
^
clang++ -DHAVE_CONFIG_H -I. -I../config    -DSYSCONFDIR="\"/usr/local/Cellar/sphinx/2.2.11/etc\"" -DDATADIR="\"/usr/local/var/data\"" -I/usr/local/include   -I../libstemmer_c/include     -fno-rtti -Wall -g -D_FILE_OFFSET_BITS=64 -O3 -DNDEBUG -c -o sphinxsearch.o sphinxsearch.cpp
1 error generated.
```

The variadic version of `call_once` is surrounded by `#ifndef _LIBCPP_HAS_NO_VARIADICS`:

```
#ifndef _LIBCPP_HAS_NO_VARIADICS

template<class _Callable, class... _Args>
_LIBCPP_INLINE_VISIBILITY
void call_once(once_flag&, _Callable&&, _Args&&...);

#else  // _LIBCPP_HAS_NO_VARIADICS

template<class _Callable>
_LIBCPP_INLINE_VISIBILITY
void call_once(once_flag&, _Callable&);

template<class _Callable>
_LIBCPP_INLINE_VISIBILITY
void call_once(once_flag&, const _Callable&);

#endif  // _LIBCPP_HAS_NO_VARIADICS
```

But that macro isn't defined anywhere.

I tried putting in `option :cxx11` and `needs :cxx11`; neither worked.

@ilovezfs Sorry for the delay; no I didn't. `ENV.cxx11` appears to trigger a different set of errors:

```
sphinxutils.cpp:2106:46: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        sphSafeInfo ( iFD, "Configured with flags: "CONFIGURE_FLAGS );
                                                    ^
                                                     
sphinxutils.cpp:2110:34: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
        sphSafeInfo ( iFD, "Host OS is "OS_UNAME );
                                        ^
                                         
2 errors generated.
make[2]: *** [sphinxutils.o] Error 1
make[2]: *** Waiting for unfinished jobs....
sphinx.cpp:27820:15: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                                                "docid="DOCID_FMT, SqlFieldName(iIndex), iPackedLen, m_tDocInfo.m_uDocID );
                                                        ^
                                                         
sphinx.cpp:27836:16: error: invalid suffix on literal; C++11 requires a space between literal and identifier [-Wreserved-user-defined-literal]
                                                " docid="DOCID_FMT, SqlFieldName(iIndex), (int)uSize, m_tDocInfo.m_uDocID );
                                                         ^
                                                          
2 errors generated.
```

This happens with a plain `brew install sphinx`, though, so I'm not sure whether it addressed the initial problem. The full gist is [here](https://gist.github.com/aw1621107/f2a743999190b4662907507b9902bf5e). The above error was in 04.make.

This one, at least, seems like an upstream bug. From what I gleaned from searches this was something that was changed between initial implementation in gcc and the final spec in the C++11 standard.

--- a/Formula/sphinx.rb
+++ b/Formula/sphinx.rb
@@ -35,7 +35,18 @@ class Sphinx < Formula
     cause "sphinxexpr.cpp:1802:11: error: use of undeclared identifier 'ExprEval'"
   end
 
+  needs :cxx11 if build.with? "re2"
+
   def install
+    if build.with? "re2"
+      ENV.cxx11
+
+      # Fix "error: invalid suffix on literal" and "error:
+      # non-constant-expression cannot be narrowed from type 'long' to 'int'"
+      # Upstream issue from 7 Dec 2016 http://sphinxsearch.com/bugs/view.php?id=2578
+      ENV.append "CXXFLAGS", "-Wno-reserved-user-defined-literal -Wno-c++11-narrowing"
+    end
+
     resource("stemmer").stage do
       system "make", "dist_libstemmer_c"
       system "tar", "xzf", "dist/libstemmer_c.tgz", "-C", buildpath

